### ISSUE TITLE
Crash: Could not allocate memory

### ISSUE THREAD

**Issue by cltnschlosser** [DataDog/dd-sdk-ios#1091]
### The crash

I have this out of memory crash, there could be a couple causes, but I think the Datadog code can be improved to help guard against this.
`Could not allocate memory.`
```
Crashed: com.datadoghq.ios-sdk-read-write
0  libswiftCore.dylib             0x37699c swift_slowAlloc.cold.1 + 16
1  libswiftCore.dylib             0x2ed944 swift_slowAlloc + 116
2  libswiftCore.dylib             0x2edac8 swift_allocObject + 52
3  libswiftCore.dylib             0xf7a0 _ContiguousArrayBuffer.init(_uninitializedCount:minimumCapacity:) + 88
4  Datadog                        0xeb434 DataBlockReader.read(length:) + 4301026356 (<compiler-generated>:4301026356)
5  Datadog                        0xeb278 DataBlockReader.readData() + 167 (DataBlock.swift:167)
6  Datadog                        0xeb330 DataBlockReader.all() + 105 (DataBlock.swift:105)
7  Datadog                        0x11e454 FileReader.decode(data:) + 63 (FileReader.swift:63)
8  Datadog                        0x11e188 FileReader.readNextBatch() + 38 (FileReader.swift:38)
9  Datadog                        0x11f094 protocol witness for Reader.readNextBatch() in conformance FileReader + 4301238420 (<compiler-generated>:4301238420)
10 Datadog                        0xffdac partial apply for closure #1 in DataReader.readNextBatch() + 4301110700 (<compiler-generated>:4301110700)
11 libswiftDispatch.dylib         0x183c partial apply for thunk for @callee_guaranteed () -> (@out A, @error @owned Error) + 20
12 libswiftDispatch.dylib         0x1854 thunk for @callee_guaranteed () -> (@out A, @error @owned Error)partial apply + 12
13 libswiftDispatch.dylib         0x29ac closure #1 in closure #1 in OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 120
14 libswiftDispatch.dylib         0x2a74 partial apply for thunk for @callee_guaranteed () -> () + 20
15 libswiftDispatch.dylib         0x2478 thunk for @escaping @callee_guaranteed () -> () + 20
16 libdispatch.dylib              0x64094 _dispatch_client_callout + 16
17 libdispatch.dylib              0x11964 _dispatch_lane_barrier_sync_invoke_and_complete + 52
18 libswiftDispatch.dylib         0x1e18 implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 152
19 libswiftDispatch.dylib         0x10ac partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 40
20 libswiftDispatch.dylib         0x1bb0 OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 256
21 libswiftDispatch.dylib         0x1140 OS_dispatch_queue.sync<A>(execute:) + 140
22 Datadog                        0xffc2c protocol witness for Reader.readNextBatch() in conformance DataReader + 4301110316 (<compiler-generated>:4301110316)
23 Datadog                        0x101c64 closure #1 in DataUploadWorker.init(queue:fileReader:dataUploader:uploadConditions:delay:featureName:) + 56 (DataUploadWorker.swift:56)
24 Datadog                        0x3eea0 thunk for @escaping @callee_guaranteed () -> () + 4300320416 (<compiler-generated>:4300320416)
25 libdispatch.dylib              0x11544 _dispatch_block_async_invoke2 + 104
26 libdispatch.dylib              0x64094 _dispatch_client_callout + 16
27 libdispatch.dylib              0x6bb8 _dispatch_continuation_pop$VARIANT$mp + 440
28 libdispatch.dylib              0x188dc _dispatch_source_invoke$VARIANT$mp + 1668
29 libdispatch.dylib              0xa610 _dispatch_lane_serial_drain$VARIANT$mp + 344
30 libdispatch.dylib              0xb1f4 _dispatch_lane_invoke$VARIANT$mp + 408
31 libdispatch.dylib              0x14ec8 _dispatch_workloop_worker_thread + 632
32 libsystem_pthread.dylib        0x1e10 _pthread_wqthread + 284
33 libsystem_pthread.dylib        0x193c start_wqthread + 8
```
```
com.datadoghq.ios-sdk-tracing-upload
0  libsystem_kernel.dylib         0x1134 kevent_id + 8
1  libdispatch.dylib              0x224b4 _dispatch_kq_poll + 228
2  libdispatch.dylib              0x22ebc _dispatch_event_loop_wait_for_ownership$VARIANT$mp + 440
3  libdispatch.dylib              0x11bdc __DISPATCH_WAIT_FOR_QUEUE__ + 288
4  libdispatch.dylib              0x11820 _dispatch_sync_f_slow + 136
5  libswiftDispatch.dylib         0x1e18 implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 152
6  libswiftDispatch.dylib         0x10ac partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 40
7  libswiftDispatch.dylib         0x1bb0 OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 256
8  libswiftDispatch.dylib         0x1140 OS_dispatch_queue.sync<A>(execute:) + 140
9  Datadog                        0xffc2c protocol witness for Reader.readNextBatch() in conformance DataReader + 4301110316 (<compiler-generated>:4301110316)
10 Datadog                        0x101c64 closure #1 in DataUploadWorker.init(queue:fileReader:dataUploader:uploadConditions:delay:featureName:) + 56 (DataUploadWorker.swift:56)
11 Datadog                        0x3eea0 thunk for @escaping @callee_guaranteed () -> () + 4300320416 (<compiler-generated>:4300320416)
12 libdispatch.dylib              0x11544 _dispatch_block_async_invoke2 + 104
13 libdispatch.dylib              0x64094 _dispatch_client_callout + 16
14 libdispatch.dylib              0x6bb8 _dispatch_continuation_pop$VARIANT$mp + 440
15 libdispatch.dylib              0x188dc _dispatch_source_invoke$VARIANT$mp + 1668
16 libdispatch.dylib              0xa610 _dispatch_lane_serial_drain$VARIANT$mp + 344
17 libdispatch.dylib              0xb1f4 _dispatch_lane_invoke$VARIANT$mp + 408
18 libdispatch.dylib              0x14ec8 _dispatch_workloop_worker_thread + 632
19 libsystem_pthread.dylib        0x1e10 _pthread_wqthread + 284
20 libsystem_pthread.dylib        0x193c start_wqthread + 8
```
```
com.datadoghq.ios-sdk-logging-upload
0  libsystem_kernel.dylib         0x1134 kevent_id + 8
1  libdispatch.dylib              0x224b4 _dispatch_kq_poll + 228
2  libdispatch.dylib              0x22ebc _dispatch_event_loop_wait_for_ownership$VARIANT$mp + 440
3  libdispatch.dylib              0x11bdc __DISPATCH_WAIT_FOR_QUEUE__ + 288
4  libdispatch.dylib              0x11820 _dispatch_sync_f_slow + 136
5  libswiftDispatch.dylib         0x1e18 implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 152
6  libswiftDispatch.dylib         0x10ac partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 40
7  libswiftDispatch.dylib         0x1bb0 OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 256
8  libswiftDispatch.dylib         0x1140 OS_dispatch_queue.sync<A>(execute:) + 140
9  Datadog                        0xffc2c protocol witness for Reader.readNextBatch() in conformance DataReader + 4301110316 (<compiler-generated>:4301110316)
10 Datadog                        0x101c64 closure #1 in DataUploadWorker.init(queue:fileReader:dataUploader:uploadConditions:delay:featureName:) + 56 (DataUploadWorker.swift:56)
11 Datadog                        0x3eea0 thunk for @escaping @callee_guaranteed () -> () + 4300320416 (<compiler-generated>:4300320416)
12 libdispatch.dylib              0x11544 _dispatch_block_async_invoke2 + 104
13 libdispatch.dylib              0x64094 _dispatch_client_callout + 16
14 libdispatch.dylib              0x6bb8 _dispatch_continuation_pop$VARIANT$mp + 440
15 libdispatch.dylib              0x188dc _dispatch_source_invoke$VARIANT$mp + 1668
16 libdispatch.dylib              0xa610 _dispatch_lane_serial_drain$VARIANT$mp + 344
17 libdispatch.dylib              0xb1f4 _dispatch_lane_invoke$VARIANT$mp + 408
18 libdispatch.dylib              0x14ec8 _dispatch_workloop_worker_thread + 632
19 libsystem_pthread.dylib        0x1e10 _pthread_wqthread + 284
20 libsystem_pthread.dylib        0x193c start_wqthread + 8
```

This leads me to 3 possibilities:
1) Datadog is generating a huge buffer.
2) Datadog buffers are being leaked somehow (or just need more tight autorelease)
3) Something in my application is using a lot of memory and Datadog is the last straw that is being hit.

---

#### Datadog SDK versions:

1.12.1

#### Last stable Datadog SDK version:

1.11.1

#### Volume:

Crash is somewhat rare (over 100 times, but only 16 users. Drop in the bucket, but still worth looking into given the poor experience. Constant crashing near startup.)

#### OS version:

iOS 15. I don't expect that to be important given it's only a few users.

#### Deployment Target:

iOS 13, iPhone

#### Device version:

Seen on: iPhone 7, iPhone 7 Plus, iPhone 8 Plus, iPhone SE

#### Environment:

Low-ish memory. <100MB, but not completely out of memory.

**Comment by ncreated** [comment]
Hey @cltnschlosser thanks for providing full context and opening a PR - we consider it valuable and correct. We need some time within the team to discuss best strategy of incorporating it.

And yes, OOMs are difficult and could come from any allocation happening in the app. I do agree that Datadog's buffer allocation should be smarter to better treat the space and to be less likely responsible for OOM in low-mem circumstances.

Could you please confirm that you're also seeing OOMs originating in other stack traces, not only Datadog? Given the repetitive nature of DD's buffers allocation it is likely to have it most often, but only to some extend.

**Comment by cltnschlosser** [comment]
I'm not seeing OOMs with other stack traces, in addition it looks like the phone still has quite a bit of memory left. I'm familiar with the phone still having like 10MB reserved for system usage, but these devices can have 50-100MB left. Also this was new to Datadog 1.12.0 SDK. So I think I've narrowed down the issue. The `DataBlock` struct is new and so is its data loading code. Prior to https://github.com/DataDog/dd-sdk-ios/pull/841 the reading of files was just a direct load to `Data`. Now the logic makes a copy of that data into a new buffer. I don't believe that the file is too big, what I think is happening is the TLV header is either corrupted or malformed resulting in an incorrect length. Causing the code to allocate a very large buffer (bigger than the data already is). The crashing line is always this: `var bytes = [UInt8](repeating: 0, count: length)`. If it were any other situation I'd expect crashes at various places (Ex: where files are loaded into memory originally).

**Comment by ncreated** [comment]
👋 Hello. We hope that with #1089 and other changes made recently on our side the situation will be improved, less likely causing OOMs in your apps. We made [some benchmarks](https://github.com/DataDog/dd-sdk-ios/issues/956#issuecomment-1384426485) comparing memory allocation among several versions of the SDK - the next release should behave visibly better. No precise ETA for `1.15.0` but I will discuss it with the team in following days.

**Comment by maxep** [comment]
Hey @cltnschlosser ! Your fix has been shipped in [`1.15.0`](https://github.com/DataDog/dd-sdk-ios/releases/tag/1.15.0), thanks again for your contribution 🙏
